olm/operator-bundle.md

```diff
@@ -0,0 +1,199 @@
+* Operator metadata can be inspected/downloaded independent of the manifests using a bundle identifier, and vice versa.
+
+The following user stories discuss the User Experience.
+
+---
+
+### User Stories
+
+#### Build, Push, Pull Operator Bundle
+As an operator author, I would like to associate operator manifests and metadata with the container image of my operator.
+
+The focus of this user story is to define a standard to store, transmit, inspect and retrieve operator manifests and metadata. The exact format of the metadata is outside of the scope of this story.
+
+Constraints:
+* An operator bundle (including both manifests and metadata) should be identifiable using a single versioned identifier. 
+* For an operator The metadata can be downloaded independently of the manifest.
+
+### Implementation Details/Notes/Constraints
+* The initial implementation target will be Docker v2-2 `manifests`, `manifest-lists`, and docker client support, for maximum compatiblity with existing tooling.
+* We want the entire operator bundle to be identifiable and retrievable using the same identifier/URL.
+
+#### Docker
+
+##### Build, Push, Pull Operator Bundle Image
+We use the following labels to annotate the operator bundle image.
+* The label `operators.coreos.com.bundle.resources` represents the bundle type:
+    * The value `manifests` implies that this bundle contains operator manifests only.
+    * The value `metadada` implies that this bundle has operator metadata only.
+    * The value `manifests+metadata` implies that this bundle contains both operator metadata and manifets.
+* The label `operators.coreos.com.bundle.mediatype` reflects the media type of the operator bundle. It could be helm charts, plain kubernetes manifests etc.
+
+The labels will also be put inside a YAML file, as shown below.
+*annotations.yaml*
+```yaml
+annotations:
+  operators.coreos.com.bundle.resources: "manifests+metadata"
+  operators.coreos.com.bundle.mediatype: bundle
+```
+
+The package manifest YAML file that defines a set of `channel`(s) will not be included in the bundle. Operator registry will provide machineries to define `package` and `channel` that can be used while index is being built. 
+
+For an operator registry bundle, the source directory structure will have similar layout as shown below.
+```
+$ tree test
+test
+├── 0.1.0
+│   ├── testbackup.crd.yaml
+│   ├── testcluster.crd.yaml
+│   ├── testoperator.v0.1.0.clusterserviceversion.yaml
+│   └── testrestore.crd.yaml
+└── annotations.yaml
+```
+
+`Dockerfile` for operator bundle
+```
+FROM scratch
+
+# We are pushing an operator-registry bundle that
+# has both metadata and manifests.
+LABEL operators.coreos.com.bundle.resources=manifests+metadata
+LABEL operators.coreos.com.bundle.mediatype=bundle
+
+ADD test/0.1.0 /manifests
+ADD test/annotations.yaml /metadata/annotations.yaml
+```
+
+Below is the directory layout of the operator bundle insdie the image.
+```bash
+$ tree
+/
+├── manifests
```

Comment:
Can you describe what you would do with this?  It doesn't look like a straight `apply` of the directory content.  It is simply free-form?

Reply:
In this example, I have used the format used by operator-registry - https://github.com/operator-framework/operator-registry#manifest-format. This is the format olm uses to package operator manifests today. The directory content in this case is not directly `kubectl apply`able. While the  operator is being installed, `olm` manipulates the content before applying it to the cluster.

I think, we can also package content in the operator bundle image that are straight `apply`.

Reply:
> In this example, I have used the format used by operator-registry - https://github.com/operator-framework/operator-registry#manifest-format. This is the format olm uses to package operator manifests today. The directory content in this case is not directly `kubectl apply`able. While the operator is being installed, `olm` manipulates the content before applying it to the cluster.
> 
> I think, we can also package content in the operator bundle image that are straight `apply`.

please include a link for reference.

Reply:
> The directory content in this case is not directly kubectl applyable

The content should be kubectl applyable, right?